### PR TITLE
CI and repo maintenance

### LINKED ISSUE #3197: Update versions of tools used in CI
The Shadow CI environment has hard-coded versions of some tools. These versions should be periodically updated, and any compatibility issues fixed.

- rust default stable version (`ci/rust-toolchain-stable.toml`)
- rust default nightly version (`ci/rust-toolchain-nightly.toml`)
- python version for the python lint (`.github/workflows/lint.yml`)
- tor and tgen versions for the tor tests (`.github/workflows/extra_tests.yml`)

Shadow's dependencies may also need to be updated.

```bash
# "cargo upgrade" won't update the lock file for all
# dependencies, so "cargo update" is also needed
cd src && cargo upgrade --incompatible && cargo update
```

*This issue is automatically generated every 4 months by the workflow "CI Maintenance".*

### PR DESCRIPTION
**Pull request by robgjansen**
- Rust was already updated in #3257
- Codecov access has now been revoked (as decided in #3143)

Closes #3197

### CODE REVIEW

File: .github/workflows/ci_maintenance.yml
Comment (stevenengler) [inline .github/workflows/ci_maintenance.yml:37]: Why is the `cargo upgrade` removed?

Comment (robgjansen) [reply .github/workflows/ci_maintenance.yml:37]: I got this:

```
$ cargo --version
cargo 1.75.0 (1d8b05cdd 2023-11-20)
$ cargo upgrade
error: no such command: `upgrade`

        Did you mean `update`?

        View all installed commands with `cargo --list`
        Find a package to install `upgrade` with `cargo search cargo-upgrade`
```

So I assumed `upgrade` was removed as a subcommand. But maybe my assumption is incorrect? What am I missing?

Comment (stevenengler) [reply .github/workflows/ci_maintenance.yml:37]: Ah sorry, I should have included that in the documentation. The `cargo upgrade` command comes from [cargo-edit](https://crates.io/crates/cargo-edit). The maintainer hopes to get some version of this merged into cargo someday.

Running `cargo update` will not update things that may have breaking changes. For example if your Cargo.toml has `foo = 1.2.0` and foo has new versions 1.2.1 and 1.3.0, `cargo update` will only upgrade the dependency to 1.2.1. If you actually want the latest versions, you need to use `cargo upgrade` to upgrade to 1.3.0.

Comment (robgjansen) [reply .github/workflows/ci_maintenance.yml:37]: OK, thank you! I've installed `cargo-edit` and will push an updated comment in `ci_maintenance.yml` to state the required package :)

**Comment by robgjansen** [comment]
@stevenengler any hints on how to fix https://github.com/shadow/shadow/actions/runs/7380020507/job/20076929685?pr=3264#step:5:328 ?

**Comment by stevenengler** [comment]
Yeah that would be a pain to fix. See #3141.

I would just manually edit `src/tests/Cargo.toml` and set the nix version back to what it was originally. You might need to also run `cargo update` again to update the lock file.

**Comment by robgjansen** [comment]
Before merging, I will remove `fedora:37` and `tor-0.4.8.7` from the required checks, and change `fedora:38` and `tor-0.4.8.10` from optional to required.

**Review by stevenengler — APPROVED** [review]

**Pull request merged**
